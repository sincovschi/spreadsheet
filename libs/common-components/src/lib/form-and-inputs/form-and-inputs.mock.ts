export type FormTypes = {
  textInput: string;
  selectInput: string;
  checkboxInput: boolean;
  hiddenInput: string;
};

export const textInputProps = {
  value: "text-value",
  label: "Text input label",
  validate: {
    required: "Required",
    minLength: {
      value: 2,
      message: "Min 2",
    },
  },
};

export const selectInputProps = {
  value: "CH",
  label: "Select input label",
  items: { CH: "Switzerland", DE: "Germany" },
  validate: {
    required: "Required",
  },
};

export const checkboxInputProps = {
  value: true,
  label: "Checkbox input label",
};

export const hiddenInputProps = {
  value: "hidden-value",
};

export const radioGroupProps = {
  options: [
    { value: "1", label: "Radio one" },
    { value: "2", label: "Radio two" },
  ],
  value: "2",
};

export const defaultValues = {
  textInput: textInputProps.value,
  selectInput: selectInputProps.value,
  checkboxInput: checkboxInputProps.value,
  radioGroup: radioGroupProps.value,
  hiddenInput: hiddenInputProps.value,
};
