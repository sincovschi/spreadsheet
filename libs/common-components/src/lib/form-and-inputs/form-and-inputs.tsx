import * as React from "react";
import {
  Box,
  Button,
  ButtonProps,
  Checkbox,
  FormControlLabel,
  FormControlLabelProps,
  Input,
  InputProps,
  MenuItem,
  Radio,
  RadioGroup,
  RadioGroupProps,
  TextField,
  TextFieldProps,
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import {
  Controller,
  DeepPartial,
  SubmitHandler,
  UnpackNestedValue,
  useForm,
  useFormContext,
  RegisterOptions,
  FormProvider,
} from "react-hook-form";
import clsx from "clsx";

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    width: "100%",
  },
}));

export function Form<T extends Record<string, unknown>>({
  onSubmit,
  defaultValues,
  className = "",
  children,
  ...props
}: {
  onSubmit: (data: T) => void;
  defaultValues?: UnpackNestedValue<DeepPartial<T>>;
  className?: string;
  children?: React.ReactChild | React.ReactChild[];
  [key: string]: unknown;
}) {
  const classes = useStyles();

  const formMethods = useForm<T>({
    mode: "onChange",
    reValidateMode: "onChange",
    defaultValues: defaultValues,
    resolver: undefined,
    criteriaMode: "firstError",
    shouldFocusError: true,
    shouldUnregister: true,
  });

  const { handleSubmit } = formMethods;

  const formSubmit = function (data: T): void {
    onSubmit(data);
  } as SubmitHandler<T>;

  return (
    <FormProvider {...formMethods}>
      <form
        className={clsx(classes.root, className)}
        onSubmit={handleSubmit(formSubmit)}
        {...props}
      >
        {children}
      </form>
    </FormProvider>
  );
}

Form.TextInput = function FormTextInput({
  name,
  label,
  className = "",
  helperText = "",
  fieldValidate = undefined,
  size = "medium",
  ...rest
}: TextFieldProps & {
  name: string;
  helperText?: string;
  fieldValidate?: RegisterOptions;
}) {
  const { register, errors } = useFormContext();

  return (
    <TextField
      name={name}
      label={label}
      className={className}
      inputRef={register(fieldValidate)}
      required={!!fieldValidate?.required}
      error={!!errors?.[name]?.message}
      helperText={errors?.[name]?.message ?? helperText}
      size={size}
      variant="outlined"
      {...rest}
    />
  );
};

Form.Select = function FormSelect({
  name,
  label,
  menuItems = {},
  helperText = "",
  fieldValidate,
  size = "medium",
  onChange: controlOnChange,
  colourSchemes,
  ...rest
}: TextFieldProps & {
  name: string;
  menuItems: Record<string, string | JSX.Element>;
  helperText?: string;
  fieldValidate?: RegisterOptions;
  colourSchemes?: unknown;
}) {
  const { control, errors } = useFormContext();

  return (
    <Controller
      name={name}
      control={control}
      rules={fieldValidate}
      render={({ name, value, onChange, onBlur }) => (
        <TextField
          select
          label={label}
          name={name}
          value={value}
          size={size}
          variant="outlined"
          {...rest}
          onChange={(event) => {
            onChange(event);
            controlOnChange?.(event);
          }}
          onBlur={onBlur}
          required={!!fieldValidate?.required}
          error={!!errors?.[name]?.message}
          helperText={errors?.[name]?.message ?? helperText}
        >
          {Object.entries(menuItems).map(([key, value]) => (
            <MenuItem key={key} value={key}>
              <Box width={1} display="flex" justifyContent="space-between">
                {value}
                <span
                  style={{
                    color: "#fff",
                    marginLeft: 10,
                    width: 100,
                    height: 20,
                  }}
                ></span>
              </Box>
            </MenuItem>
          ))}
        </TextField>
      )}
    />
  );
};

Form.CheckboxInput = function FormCheckboxInput({
  name,
  label,
  disabled = false,
  size = "medium",
  ...rest
}: Omit<FormControlLabelProps, "control"> & {
  name: string;
  size?: "small" | "medium";
}) {
  const { register, control } = useFormContext();

  return (
    <FormControlLabel
      label={label}
      {...rest}
      control={
        <Checkbox
          defaultChecked={!!control.defaultValuesRef.current[name]}
          name={name}
          size={size}
          disabled={disabled}
          inputRef={register()}
          color="primary"
        />
      }
    />
  );
};

Form.Radio = function FormRadio({
  name,
  value,
  label,
  disabled = false,
  size = "medium",
  onChange = () => undefined,
  ...rest
}: Omit<FormControlLabelProps, "control"> & {
  name: string;
  value: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  size?: "small" | "medium";
}) {
  const { register, watch, setValue } = useFormContext();

  const stringValue = String(value);
  const checkedValue = watch(name);

  return (
    <FormControlLabel
      label={label}
      {...rest}
      control={
        <Radio
          checked={stringValue === checkedValue}
          onChange={(event) => {
            setValue(name, stringValue);
            onChange(event);
          }}
          inputRef={register()}
          name={name}
          value={value}
          size={size}
          disabled={disabled}
          color="primary"
        />
      }
    />
  );
};

Form.RadioGroup = function FormRadioGroup({
  name,
  options,
  size = "medium",
  onChange = () => undefined,
  ...rest
}: RadioGroupProps & {
  name: string;
  options: { value: string; label: string; disabled?: boolean }[];
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  size?: "small" | "medium";
}) {
  const radioOptions = options.map((opt) => {
    const { value, label, disabled } = opt;
    return (
      <Form.Radio
        key={value}
        name={name}
        label={label}
        value={value}
        disabled={disabled}
        size={size}
      />
    );
  });

  return (
    <RadioGroup onChange={onChange} name={name} {...rest}>
      {radioOptions}
    </RadioGroup>
  );
};

Form.Button = function FormButton({
  label,
  type = "button",
  size = "medium",
  variant = "contained",
  color = "primary",
  disableOnInvalidForm = false,
  ...rest
}: ButtonProps & {
  label: string;
  disableOnInvalidForm?: boolean;
}) {
  const { formState } = useFormContext();

  return (
    <Button
      type={type}
      size={size}
      variant={variant}
      color={color}
      disabled={disableOnInvalidForm && !formState.isValid}
      {...rest}
    >
      {label}
    </Button>
  );
};

Form.HiddenInput = function FormHiddenInput({
  name,
  value = "",
  ...rest
}: InputProps & {
  name: string;
  value?: string;
}) {
  const { register, watch } = useFormContext();
  const watchedValue = watch(name);

  return (
    <Input
      name={name}
      value={watchedValue ?? value}
      {...rest}
      inputRef={register()}
      type={"hidden"}
    />
  );
};
