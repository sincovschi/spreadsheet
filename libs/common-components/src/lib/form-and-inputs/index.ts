export * from "./form-and-inputs";
