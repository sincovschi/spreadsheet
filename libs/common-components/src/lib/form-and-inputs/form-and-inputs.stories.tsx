import React from "react";
import { Form } from "./form-and-inputs";
import {
  FormTypes,
  defaultValues,
  textInputProps,
  selectInputProps,
  checkboxInputProps,
  radioGroupProps,
} from "./form-and-inputs.mock";
import { makeStyles, Theme } from "@material-ui/core";

export default {
  title: "Form and inputs component",
};

const useStyles = makeStyles<Theme>((theme) => ({
  formInput: {
    margin: theme.spacing(0, 3, 2),
  },
}));

export function FormAndInputs() {
  const classes = useStyles();

  return (
    <Form<FormTypes>
      onSubmit={console.log}
      defaultValues={defaultValues}
      style={{ width: "375px" }}
    >
      <Form.TextInput
        id={"text-input"}
        name={"textInput"}
        label={textInputProps.label}
        fieldValidate={textInputProps.validate}
      />

      <br />

      <Form.Select
        id={"select-input"}
        name={"selectInput"}
        label={selectInputProps.label}
        menuItems={selectInputProps.items}
        fieldValidate={selectInputProps.validate}
      />

      <Form.CheckboxInput
        id={"checkbox-input"}
        name={"checkboxInput"}
        label={checkboxInputProps.label}
      />

      <Form.RadioGroup name={"radioGroup"} options={radioGroupProps.options} />

      <Form.HiddenInput name={"hiddenInput"} />

      <Form.Button
        label={"Submit"}
        type={"submit"}
        disableOnInvalidForm={true}
      />

      <Form.Button
        className={classes.formInput}
        label={"Back"}
        variant={"text"}
        onClick={() => console.log("Go back.")}
      />
    </Form>
  );
}
