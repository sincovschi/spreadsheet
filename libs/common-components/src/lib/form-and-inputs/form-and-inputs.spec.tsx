import React from "react";
import { Form } from "./form-and-inputs";
import {
  FormTypes,
  defaultValues,
  textInputProps,
  selectInputProps,
  checkboxInputProps,
  radioGroupProps,
} from "./form-and-inputs.mock";
import { render, screen, waitFor, act } from "@testing-library/react";
import { default as userEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";

const formJsx = (
  <Form<FormTypes>
    onSubmit={console.log}
    defaultValues={defaultValues}
    name={"testForm"}
  >
    <Form.TextInput
      id={"text-input"}
      name={"textInput"}
      label={textInputProps.label}
      fieldValidate={textInputProps.validate}
    />
    <Form.Select
      id={"select-input"}
      name={"selectInput"}
      label={selectInputProps.label}
      menuItems={selectInputProps.items}
      fieldValidate={selectInputProps.validate}
    />
    <Form.CheckboxInput
      id={"checkbox-input"}
      name={"checkboxInput"}
      label={checkboxInputProps.label}
    />
    <Form.RadioGroup name={"radioGroup"} options={radioGroupProps.options} />
    <Form.HiddenInput name={"hiddenInput"} />
    <Form.Button label={"Submit"} type={"submit"} disableOnInvalidForm={true} />
  </Form>
);

const getForm = (doc: HTMLDocument) => {
  const form = Array.from(doc.forms).find(
    (f: HTMLFormElement) => f.name === "testForm",
  );
  if (!form) throw new Error("testForm not found!");
  return form;
};

/**
 * Generate RegExp from a string
 * with options ignore case and multiline
 * @param str
 */
const rex = (str: string) => new RegExp(str, "i");

describe("Form", () => {
  beforeEach(() => {
    render(formJsx);
  });

  it("should render and match snapshot", () => {
    expect(screen.getByRole("form")).toMatchSnapshot();
  });

  it("TextInput should have default value and validation", async () => {
    const getTextInput = () => screen.getByLabelText(rex(textInputProps.label));

    expect(getTextInput()).toHaveValue(textInputProps.value);

    await waitFor(async () => {
      userEvent.type(getTextInput(), "{selectall}{del}");
      expect(getTextInput()).toHaveValue("");
      expect(
        screen.findByText(rex(textInputProps.validate.required)),
      ).toBeTruthy();
    });

    await waitFor(async () => {
      userEvent.type(getTextInput(), "a");
      expect(getTextInput()).toHaveValue("a");
      expect(
        screen.findByText(rex(textInputProps.validate.minLength.message)),
      ).toBeTruthy();
    });
  });

  it("Select should have default value and toggle list", async () => {
    const form = getForm(document);
    expect(form).toBeTruthy();

    const getSelectVisualInput = () =>
      screen.getByLabelText(rex(selectInputProps.label));
    const getSelectValueInput = () =>
      form.querySelector<HTMLInputElement>("input[name=selectInput]");

    expect(getSelectVisualInput()).toHaveTextContent(
      selectInputProps.items[selectInputProps.value as "CH" | "DE"],
    );
    expect(getSelectValueInput()).toHaveValue(defaultValues.selectInput);

    userEvent.click(getSelectVisualInput());
    expect(screen.getByRole("listbox")).toBeTruthy();

    const options = screen.getAllByRole("option");
    expect(options[1]).toContainHTML(selectInputProps.items.DE);
    userEvent.click(options[1]);
    expect(screen.queryByRole("listbox")).toBeFalsy();

    await waitFor(async () => {
      // Needs to be waited because dom gets re-rendered
      // and exist multiple copies of same element
      expect(getSelectVisualInput()).toHaveTextContent(
        selectInputProps.items.DE,
      );
      expect(getSelectValueInput()).toHaveValue("DE");
    });
  });

  it("CheckboxInput should have default value and toggle checks", async () => {
    const form = getForm(document);
    expect(form).toBeTruthy();
    const getCheckboxLabel = () => screen.getByText(checkboxInputProps.label);
    const getCheckboxInput = () =>
      form.querySelector<HTMLInputElement>("input[name=checkboxInput]");
    expect(getCheckboxLabel()).toBeVisible();

    userEvent.click(getCheckboxInput());
    expect(getCheckboxInput()).toHaveProperty(
      "checked",
      !defaultValues.checkboxInput,
    );
  });

  it("RadioGroup should have default value and switch option", async () => {
    const form = getForm(document);
    expect(form).toBeTruthy();

    const getRadio2 = () =>
      screen.getByLabelText(rex(radioGroupProps.options[1].label));

    expect(getRadio2()).toHaveProperty(
      "checked",
      defaultValues.radioGroup === radioGroupProps.options[1].value,
    );

    const getRadio1 = () =>
      screen.getByLabelText(rex(radioGroupProps.options[0].label));

    userEvent.click(getRadio1());
    // MUI Radio reacts only to .click()
    getRadio1().click();

    await act(async () => {
      await expect(getRadio1()).toHaveProperty("checked", true);
    });
  });

  it("HiddenInput should have default value", () => {
    const form = getForm(document);
    expect(form).toBeTruthy();

    const hiddenInput = form.querySelector<HTMLInputElement>(
      "input[name=hiddenInput]",
    );
    expect(hiddenInput).toHaveValue(defaultValues.hiddenInput);
  });

  it("Button should have label and is disabled on invalid form", async () => {
    const form = getForm(document);
    expect(form).toBeTruthy();

    const getSubmitButton = () =>
      screen.getByRole("button", { name: /Submit/i });

    const getTextInput = () => screen.getByLabelText(rex(textInputProps.label));

    expect(getSubmitButton()).toBeEnabled();

    await waitFor(() => {
      userEvent.type(getTextInput(), "{selectall}{del}");
    });
    expect(getSubmitButton()).toBeDisabled();
  });
});
