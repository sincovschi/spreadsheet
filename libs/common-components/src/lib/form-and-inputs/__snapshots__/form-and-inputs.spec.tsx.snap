// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render and match snapshot 1`] = `
<form
  class="makeStyles-root-1"
  name="testForm"
>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="text-input"
      id="text-input-label"
    >
      Text input label
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input"
        id="text-input"
        name="textInput"
        required=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-4 PrivateNotchedOutline-legendNotched-5"
        >
          <span>
            Text input label
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
      data-shrink="true"
      for="select-input"
      id="select-input-label"
    >
      Select input label
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="select-input-label select-input"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
        id="select-input"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          Switzerland
          <span
            style="color: rgb(255, 255, 255); margin-left: 10px; width: 100px; height: 20px;"
          />
        </div>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="selectInput"
        required=""
        tabindex="-1"
        value="CH"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-4 PrivateNotchedOutline-legendNotched-5"
        >
          <span>
            Select input label
             *
          </span>
        </legend>
      </fieldset>
    </div>
    
  </div>
  <label
    class="MuiFormControlLabel-root"
    id="checkbox-input"
  >
    <span
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-8 Mui-checked MuiIconButton-colorPrimary"
    >
      <span
        class="MuiIconButton-label"
      >
        <input
          checked=""
          class="PrivateSwitchBase-input-10"
          data-indeterminate="false"
          name="checkboxInput"
          type="checkbox"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </span>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Checkbox input label
    </span>
  </label>
  <div
    class="MuiFormGroup-root"
    role="radiogroup"
  >
    <label
      class="MuiFormControlLabel-root"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorPrimary MuiIconButton-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-10"
            name="radioGroup"
            type="radio"
            value="1"
          />
          <div
            class="PrivateRadioButtonIcon-root-11"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        Radio one
      </span>
    </label>
    <label
      class="MuiFormControlLabel-root"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-checked-8 Mui-checked MuiIconButton-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input-10"
            name="radioGroup"
            type="radio"
            value="2"
          />
          <div
            class="PrivateRadioButtonIcon-root-11 PrivateRadioButtonIcon-checked-13"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-12"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        Radio two
      </span>
    </label>
  </div>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline"
  >
    <input
      class="MuiInputBase-input MuiInput-input"
      name="hiddenInput"
      type="hidden"
      value="hidden-value"
    />
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;
